### PR TITLE
PR: More namespace fixes

### PR DESCRIPTION
**Pull request by impact27**
A function definition was not correctly capturing the locals. Fixed and added a regression test.

### CODE REVIEW

File: spyder_kernels/customize/spyderpdb.py [hunk 1]
```diff
@@ -204,29 +186,101 @@ def default(self, line):
...
-                    #    one.
-                    # 2. Any edit to that variable will be lost.
-                    # 3. The globals will appear to contain all the locals
-                    #    variables.
-                    # 4. Any new locals variable will be saved to globals
-                    #    instead
-                    fake_globals = globals.copy()
-                    fake_globals.update(locals)
-                    locals_keys = locals.keys()
+                # If this is ever fixed in python, this whole block can be
```
Comment (ccordoba12) [inline spyder_kernels/customize/spyderpdb.py:189]:
```suggestion
                # If this is ever fixed in Python, this whole block can be
```

File: spyder_kernels/customize/spyderpdb.py [hunk 2]
```diff
@@ -204,29 +186,101 @@ def default(self, line):
...
+                # Replaced by `exec(code, globals, locals)` (else included)
+                if locals is not globals:
+                    # frame.f_globals is a mix of the "real" globals and
+                    # locals. `globals()` returns a copy of the real globals.
+                    # frame.f_globals is None. `locals()` returns the real
+                    # locals
```
Comment (ccordoba12) [inline spyder_kernels/customize/spyderpdb.py:195]:
```suggestion
                    # locals.
```

File: spyder_kernels/customize/spyderpdb.py [hunk 3]
```diff
@@ -204,29 +186,101 @@ def default(self, line):
...
+                    #    lost, except additions.
+                    # 3. frame.f_globals and globals() do not match
+                    # 4. frame.f_locals and locals() do not match
+                    #
+                    # The last two are necessary to fix the issues above.
+                    # The first two are a consequence of not being able to
+                    # detect if a variable was modified in a dict.
+                    # The solution of subclassing dict does not work as `exec`
+                    # For example accesses directly the c object, bypassing
+                    # any python code. If a way is found of detecting
```
Comment (ccordoba12) [inline spyder_kernels/customize/spyderpdb.py:209]:
```suggestion
                    # any Python code. If a way is found of detecting
```

File: spyder_kernels/customize/spyderpdb.py [hunk 4]
```diff
@@ -204,29 +186,101 @@ def default(self, line):
...
+                    # modifications after the exec, ordering becomes an issue.
+                    #
+                    # As long as the user does not modify explicitrely
```
Comment (ccordoba12) [inline spyder_kernels/customize/spyderpdb.py:212]:
```suggestion
                    # As long as the user does not modify explicitly
```

File: spyder_kernels/customize/spyderpdb.py [hunk 5]
```diff
@@ -204,29 +186,101 @@ def default(self, line):
...
+                    # frame.f_globals or frame.f_locals,
+                    # this should work as expected.
+
+                    # Save original state
+                    original_locals_keys = list(locals.keys())
+                    original_globals_keys = list(globals.keys())
+
+                    # Save a copy of the globals for new_globals
+                    globals_copy = globals.copy()
+                    # Mix the locals with the globals
```
Comment (ccordoba12) [inline spyder_kernels/customize/spyderpdb.py:223]:
```suggestion

                    # Mix the locals with the globals
```

File: spyder_kernels/customize/spyderpdb.py [hunk 6]
```diff
@@ -204,29 +186,101 @@ def default(self, line):
...
+                    globals.update(locals)
+
+                    # Create new `locals` and `globals` function that return
+                    # the right dictionnary
+
```
Comment (ccordoba12) [inline spyder_kernels/customize/spyderpdb.py:228]:
```suggestion
                    # Create new `locals` and `globals` functions that return
                    # the right dictionaries.
```

File: spyder_kernels/customize/spyderpdb.py [hunk 7]
```diff
@@ -204,29 +186,101 @@ def default(self, line):
...
+                    builtins.globals = globals_func
+                    builtins.locals = locals_func
+
+                    # Addition to `globals()`
+                    for key in globals_copy:
+                        if (key not in original_globals_keys
+                                and key not in globals):
+                            globals[key] = globals_copy[key]
+
+                    # put locals back in locals, and unmask variables
```
Comment (ccordoba12) [inline spyder_kernels/customize/spyderpdb.py:268]:
```suggestion
                    # Put locals back in locals, and unmask variables
```

**Review by ccordoba12 — COMMENTED** [review]
Thanks @impact27 for your continued work on this! I left some minor suggestions, otherwise looks good to me.

**Comment by impact27** [comment]
I opened https://bugs.python.org/issue46153

**Review by ccordoba12 — APPROVED** [review]
Thanks @impact27!

**Pull request merged**